Testing varnish config

procrastinatio opened the pull request:
This PR defines the ressources that should be "protected" by varnish.

To test varnish configuration for production:

     ./run_nose.sh -p -a

with `-a`for all tests (varnish and mapproxy are deactivated by default) and `-p` for using the procution settings.

procrastinatio commented:
This needs to be hosted or used a proxy outside the KOMBV, because of the BIT Proxy IP.

gjn commented:
We can ssh into ssh0a, copy the relevant stuff and run the tests from there. I think the testscript(s) of our test project should take care of that for the varnish tests (probably all e2e tests should be run on both sides).

procrastinatio commented:
I leave the test here, but use a custom `http_proxy` for the requests testing varnish. I'll update the `http_proxy` as soon as I'll get the definitive address.

procrastinatio commented:
Using the new frankfurter proxy: ec2-54-93-109-29.eu-central-1.compute.amazonaws.com:80

gjn commented on chsdi/tests/mapproxy/test_wmtscapabilities.py:
```diff
@@ -0,0 +1,112 @@
+import random
+from unittest import TestCase
+from testconfig import config as tc
+
```
`import requests` is missing here, meaning you can't run chsdi/tests/mapproxy/test_wmtscapabilities.py alone.

gjn commented on chsdi/tests/mapproxy/test_wmtscapabilities.py:
```diff
@@ -0,0 +1,112 @@
+
+    def setUp(self):
+        try:
+            self.url = "http:" + tc['vars']['mapproxy_url']
```
`mapproxy_url` is nowhere defined. Did you mean to use `mapproxyhost` here?

gjn commented on chsdi/tests/mapproxy/test_wmtscapabilities.py:
```diff
@@ -0,0 +1,112 @@
+        try:
+            self.url = "http:" + tc['vars']['mapproxy_url']
+        except KeyError as e:
+            self.url = os.getenv('conf_to_use', "http://api3.geo.admin.ch")
```
The name 'conf_to_use' is confusing to me. Why not name it mapproxyhost? Or wmts_url?

gjn commented on chsdi/tests/mapproxy/test_wmtscapabilities.py:
```diff
@@ -0,0 +1,112 @@
+            resp = requests.get(self.url + '/rest/services/api/1.0.0/WMTSCapabilities.xml', params={'epsg': epsg, '_id': self.hash()},
+                                headers=self.get_headers())
+
+            print resp.encoding
```
This should be removed.

gjn commented on chsdi/tests/mapproxy/test_wmtsgettile.py:
```diff
@@ -0,0 +1,86 @@
+
+
+def check_status_code(path):
+    print path
```
Should be removed.

gjn commented on chsdi/tests/varnish/test_varnish.py:
```diff
@@ -0,0 +1,239 @@
+        return random.randrange(20140101, 20141001)
+
+    def setUp(self):
+        os.environ["http_proxy"] = tc['vars']['http_proxy']
```
Can't we set a default proxy? So we could run it without loading a configuration?

e.g. running `buildout/bin/nosetests chsdi/tests/varnish/test_varnish.py' should also work, no?

gjn commented:
A few comments:

1) For such e2e tests, why not just use the official urls, as published in www.geo.admin.ch and http://api3.geo.admin.ch/services/sdiservices.html#wmts? If those are ok, we know backend/mapproxy are ok, no?
2) I couldn't run `chsdi/tests/mapproxy/test_wmtsgettile.py` succesfully. Had 306 failures.
3) For the tiles, shouldn't we mix the official adresses wmts, wmts1, wmts2, etc...(because that's what we need to support)?
4) We should be able to exclude those tests from standard nosetest runs (without any parameters, or specify to ignore them) as they are to be executed _after_ deploy. All other nosetests can be run _before_ the deploy. Then we need to adapt the deploy scripts accordingly (and also jenkis integration). If you don't mind, I'll create a PR to this branch for this.

gjn commented:
jenkins test

procrastinatio commented:
@gjn 

1) We test the varnish settings also on _dev_ and _int_ before _prod_, but _prod_ is by far more complicated than the other two (because of aliases). But why not.

2) The WMTS Capabilities reads the BOD, Mapproxy uses a generated config from the BOD. If both do not match, you have errors.

3) OK. Plus the various aliases.

4) For Varnish, sure it has to be done after deployment, and after every varnish changes. For WMTS, ideally it should be run **before**, but because of the staging mess, it has to be run **after**, for check.

gjn commented:
@procrastinatio I've create a PR to address parts of point 4. See https://github.com/geoadmin/mf-chsdi3/pull/1314

We can discuss the rest tomorrow. I want this in!

gjn commented on chsdi/tests/mapproxy/test_wmtscapabilities.py:
```diff
@@ -0,0 +1,112 @@
+             }
+
+
+class TestWmtsCapabilitiesView(TestCase):
```
I think this could be moved into the 'integration' tests (https://github.com/geoadmin/mf-chsdi3/blob/master/chsdi/tests/integration/test_wmtscapabilities.py#L6) as we don't test anything varnish related. It's fast as well, so it can be run on standard nose tests.

procrastinatio replied:
OK, moved to *integrations* tests.

gjn commented on chsdi/tests/mapproxy/test_wmtscapabilities.py:
```diff
@@ -0,0 +1,112 @@
+            resp = requests.get(url, params={'_id': self.hash()})
+        assert int(resp.status_code) == code
+
+    def test_bad_referer_get_capabilties(self):
```
Use address official (without rest/...)

procrastinatio commented:
Updated according to comments/discussion:
* Using the only `run_nose.sh` with a supplementry parameter `-a` (for all tests)
* Moved some test to `integration`
* Testing only the official Urls

procrastinatio commented:
jenkins test

gjn commented:
Awewome I'll merge. Thanks alot for this much needed effort.

@daguer @alcapat @ponceta We will use these new tests on our ci instance.
